ext/spl/spl_directory.c

```diff
@@ -2046,8 +2046,13 @@ static int spl_filesystem_file_read(spl_filesystem_object *intern, int silent) /
 		intern->u.file.current_line_len = 0;
 	} else {
 		if (SPL_HAS_FLAG(intern->flags, SPL_FILE_OBJECT_DROP_NEW_LINE)) {
-			line_len = strcspn(buf, "\r\n");
-			buf[line_len] = '\0';
+			zend_string *zbuf = zend_string_init(buf, line_len, 0);
```

Comment:
Going through a temporary zend_string here seems rather inefficient. Can't we do something like strip \r/\n from the end of buf?

Reply:
Don't we need to heed the EOL setting of the stream, i.e. what [`php_stream_locate_eol`](https://heap.space/xref/PHP-7.4/main/streams/streams.c?r=963e50c8#850) does?

Reply:
Don't we implicitly do that because we know that we must end on a newline, so it's okay to strip any of \n, \r or \r\n?

... Or maybe not, I guess you could have the situation where you have a `\r` at the very end of the file and it should not count as a newline in that case?

This probably needs some more tests, with and without auto_detect_line_endings, and with the wrong newline kind at the end of the string.

Reply:
Seeing what auto_detect_line_endings actually does, I added a stub for it to https://wiki.php.net/rfc/deprecations_php_8_1 -- doesn't make sense to support this anymore.

I'm happy to sign off on this if you ignore auto_detect_line_endings and simply do not support isolated `\r` entirely here.

Reply:
(Though if you do want to be pedantic, just checking the EOL_MAC flag should be sufficient at this point, as EOL detection should have already happened.)

Reply:
> I guess you could have the situation where you have a \r at the very end of the file and it should not count as a newline in that case?

Yes; that is already tested, but I can add more tests for other scenarios.

Reply:
Oh, I missed your two latest comments. I have now changed to only detect CRLF and LF.